engine/src/main/java/org/hibernate/validator/cfg/AnnotationDef.java

```diff
@@ -64,12 +64,16 @@
 	 */
 	private final Map<String, Class<?>> annotationsAsParametersTypes;
 
-	protected AnnotationDef(Class<A> annotationType) {
-		this.annotationDescriptorBuilder = new AnnotationDescriptor.Builder<>( annotationType );
+	protected AnnotationDef(AnnotationDescriptor.Builder<A> annotationDescriptorBuilder) {
+		this.annotationDescriptorBuilder = annotationDescriptorBuilder;
```

Comment:
Unfortunately, we can't do that as we would expose internal classes in a public facing API.

We have the same issue with the `protected` trick we used to expose `createAnnotation()` without making it public. With `createAnnotationDescriptor()`, it does not work anymore as it exposes `AnnotationDescriptor` to the subclasses. We created https://hibernate.atlassian.net/browse/HV-1511 for this one (maybe you could fix it? Basically, it's about making `createAnnotationDescriptor()` private and calling it via reflection - with the usual `run()` method to check the security manager).

I think the only thing you can do is having the ability to build a `ConstaintAnnotationDescriptor.Builder` from an `AnnotationDescriptor.Builder`. The fact is that you don't use your `withMessage()` and al. methods in here so it shouldn't be a big change.

I think you can keep the `withMessage()` and al. methods even if they are only useful in the tests. I think I would call them `setMessage()` though as they are basically setters. What was your reasoning for using `with...()`?

Reply:
yes... thought that might not be a good idea to change the visibility :) as for the methods they are used in the `ConstraintDescriptorImpl` as well. 
>What was your reasoning for using with...()?

I picked it up somewhere and start to call those builder methods `with*` :smiley: I'll change those to `set*` no problem. I'll see about that HV-1511 as well a bit later today.

Reply:
It's not an issue in `ConstraintDescriptorImpl` as it's an internal class and they are not exposed in the public API.